### PR TITLE
For search function adding a precision control

### LINKED ISSUE #8573: For search function adding a precision control
#### What would you like to be added:
for search function in pymilvus adding a parameter to implement a precision control , for example if parameter added "round_decimal" is 3 , the output will be 3.145 compared with current 3.14523432.
#### Why is this needed:
some precision is not necessary and in C++  float is 32bit while in python that is presented with 64bit. This difference would make precision loss.

### PR DESCRIPTION
**Pull request by elfisworking**
This PR add a “round_decimal” parameter for search function which helps search function to implement precision control(e.g. when round_decimal is 3, original output 3.134234 will be 3.134)

Resolves: #8573

Signed-off-by: elfisworking zymustb@126.com

### CODE REVIEW

**Comment by elfisworking** [comment]
add issue: https://github.com/milvus-io/milvus/issues/8573

**Comment by elfisworking** [comment]
test image when round_decimal that helps control precision is 3

**Comment by elfisworking** [comment]
test image when round_decimal that helps control precision is 3

**Comment by elfisworking** [comment]
related pymilvus PR: https://github.com/milvus-io/pymilvus/pull/733

**Comment by yanliang567** [comment]
so far i think you shall update this pr:
1. some untracked files are in the commit
2. one pr shall only has one commit

**Review by xiaocai2333 — COMMENTED** [review]
coverage.txt should not be tracked.

File: go.mod [hunk 1]
```diff
@@ -29,6 +29,7 @@ require (
 	github.com/pierrec/lz4 v2.5.2+incompatible // indirect
 	github.com/pkg/errors v0.9.1
 	github.com/prometheus/client_golang v1.11.0
+	github.com/quasilyte/go-ruleguard v0.2.1 // indirect
```
Comment (xiaocai2333) [inline go.mod:32]: Should be ignore.

File: go.mod [hunk 2]
```diff
@@ -53,4 +54,5 @@ replace (
 	github.com/apache/pulsar-client-go => github.com/apache/pulsar-client-go v0.6.0
 	google.golang.org/grpc => google.golang.org/grpc v1.38.0
 	github.com/keybase/go-keychain => github.com/99designs/go-keychain v0.0.0-20191008050251-8e49817e8af4
+	google.golang.org/grpc => google.golang.org/grpc v1.38.0
```
Comment (xiaocai2333) [inline go.mod:57]: Should be ignore too.

File: go.sum [hunk 3]
```diff
@@ -435,6 +435,8 @@ github.com/prometheus/procfs v0.6.0 h1:mxy4L2jP6qMonqmq+aTtOx1ifVWUgG/TAmntgbh3x
 github.com/prometheus/procfs v0.6.0/go.mod h1:cz+aTbrPOrUb4q7XlbU9ygM+/jj0fzG6c1xBZuNvfVA=
 github.com/prometheus/tsdb v0.7.1/go.mod h1:qhTCs0VvXwvX/y3TZrWD7rabWM+ijKTux40TwIPHuXU=
 github.com/protocolbuffers/protobuf v3.17.3+incompatible h1:weIpdqbAakIy/7PnlmdSBnPdODTtUySpnf3LyypYPwA=
+github.com/quasilyte/go-ruleguard v0.2.1 h1:56eRm0daAyny9UhJnmtJW/UyLZQusukBAB8oT8AHKHo=
+github.com/quasilyte/go-ruleguard v0.2.1/go.mod h1:hN2rVc/uS4bQhQKTio2XaSJSafJwqBUWWwtssT3cQmc=
```
Comment (xiaocai2333) [inline go.sum:439]: Same as above, should be ignore.

File: go.sum [hunk 4]
```diff
@@ -811,6 +813,7 @@ golang.org/x/tools v0.0.0-20200618134242-20370b0cb4b2/go.mod h1:EkVYQZoAsY45+roY
 golang.org/x/tools v0.0.0-20200619180055-7c47624df98f/go.mod h1:EkVYQZoAsY45+roYkvgYkIh4xh/qjgUK9TdY2XT94GE=
 golang.org/x/tools v0.0.0-20200729194436-6467de6f59a7/go.mod h1:njjCfa9FT2d7l9Bc6FUM5FLjQPp3cFF28FI3qnDFljA=
 golang.org/x/tools v0.0.0-20200804011535-6c149bb5ef0d/go.mod h1:njjCfa9FT2d7l9Bc6FUM5FLjQPp3cFF28FI3qnDFljA=
+golang.org/x/tools v0.0.0-20200812195022-5ae4c3c160a0/go.mod h1:njjCfa9FT2d7l9Bc6FUM5FLjQPp3cFF28FI3qnDFljA=
```
Comment (xiaocai2333) [inline go.sum:816]: And remove this.

**Comment by elfisworking** [comment]
ok , i will fix these problems

**Comment by elfisworking** [comment]
> so far i think you shall update this pr:
> 
> 1. some untracked files are in the commit
> 2. one pr shall only has one commit

i will fix these problems

File: go.mod (hunk 1 above)
Comment (elfisworking) [reply go.mod:32]: remove green code, right?

**Comment by xiaofan-luan** [comment]
@elfisworking hi elfis, welcome to the community! please take a look about how to do DCO, there is a guide about this https://github.com/milvus-io/milvus/blob/master/CONTRIBUTING.md

**Comment by elfisworking** [comment]
> @elfisworking hi elfis, welcome to the community! please take a look about how to do DCO, there is a guide about this https://github.com/milvus-io/milvus/blob/master/CONTRIBUTING.md

thanks your help. I have realized this problem.  Before this pr , i did  a stupid merge operation, so i has difficulty in doing git rebase operation. I will try to fixed it, maybe tomorrow would be OK.

File: internal/querynode/segment.go [hunk 5]
```diff
@@ -295,7 +295,7 @@ func (s *Segment) search(plan *SearchPlan,
 	log.Debug("do search on segment", zap.Int64("segmentID", s.segmentID), zap.Int32("segmentType", int32(s.segmentType)))
 	var status = C.Search(s.segmentPtr, plan.cSearchPlan, cPlaceHolderGroup, ts, &searchResult.cSearchResult)
 	errorCode := status.error_code
-
+	// log.Debug("do search on segment result is ",zap.Any("searchResult.cSearchResult one distance value",searchResult))
```
Comment (xiaocai2333) [inline internal/querynode/segment.go:298]: Delete the commented out code, or open it.

File: internal/querynode/segment_test.go [hunk 6]
```diff
@@ -492,8 +492,8 @@ func TestSegment_segmentSearch(t *testing.T) {
 	err = segment.segmentInsert(offset, &ids, &timestamps, &records)
 	assert.NoError(t, err)
 
-	dslString := "{\"bool\": { \n\"vector\": {\n \"vec\": {\n \"metric_type\": \"L2\", \n \"params\": {\n \"nprobe\": 10 \n},\n \"query\": \"$0\",\"topk\": 10 \n } \n } \n } \n }"
-
+	// dslString := "{\"bool\": { \n\"vector\": {\n \"vec\": {\n \"metric_type\": \"L2\", \n \"params\": {\n \"nprobe\": 10 \n},\n \"query\": \"$0\",\"topk\": 10 \n } \n } \n } \n }"
```
Comment (xiaocai2333) [inline internal/querynode/segment_test.go:495]: Delete the commented out code.

**Comment by claireyuw** [comment]
I reckon that when this PR is merged we should update the documentation to include this new parameter. @LocoRichard

File: go.mod (hunk 1 above)
Comment (xiaocai2333) [reply go.mod]: yes

**Comment by scsven** [comment]
@elfisworking Thanks for your contributions!

/kind enhancement
/approve

**Comment by claireyuw** [comment]
@elfisworking @xiaofan-luan Can you confirm that this newly added parameter is only applicable in Python and Go, not in Node.js yet? Once confirmed, @LocoRichard will update the Search docs accordingly.

**Pull request merged**